cpp/arcticdb/python/python_handlers.cpp

```diff
@@ -149,90 +188,193 @@ int BoolHandler::type_size() const {
-        auto block_pos = 0u;
-        const auto *ptr_src = (*block_it)->data();
-        constexpr stride_t stride = static_cast<TypeDescriptor>(tdt).get_type_byte_size();
-        for (auto en = bv->first(); en < bv->end(); ++en) {
-            const shape_t shape = shapes ? *shapes : 0;
-            const auto offset = *en;
-            ptr_dest = fill_with_none(ptr_dest, offset - last_row);
-            last_row = offset;
-            *ptr_dest++ = initialize_array(py_dtype,
-                                           shape,
-                                           stride,
-                                           ptr_src + block_pos,
-                                           column,
-                                           initialize_array_mutex);
-            block_pos += shape * stride;
-            if (shapes) {
-                ++shapes;
-            }
-            if (block_it != blocks.end() && block_pos == (*block_it)->bytes() && ++block_it != blocks.end()) {
-                ptr_src = (*block_it)->data();
-                block_pos = 0;
-            }
+    py::gil_scoped_acquire acquire_gil;
+    const auto &sparse_map = source_column.opt_sparse_map();
+    const auto strides = static_cast<stride_t>(get_type_size(source_type_desc.data_type()));
+    py::dtype py_dtype = generate_python_dtype(source_type_desc, strides);
 
-            ++last_row;
-        }
-        if (block_it != blocks.end() && block_pos == (*block_it)->bytes() && ++block_it != blocks.end()) {
-            ptr_src = (*block_it)->data();
-            block_pos = 0;
-        }
+    if (source_column.empty())
+        return;
+
+    auto column_data = source_column.data();
+    if (source_column.is_sparse()) {
+        auto& handler_data = get_handler_data(any);
+        python_util::prefill_with_none(ptr_dest, num_rows, source_column.sparse_map().count(), handler_data.spin_lock());
 
-        ++last_row;
-    });
+        auto en = sparse_map->first();
 
-    ARCTICDB_SUBSAMPLE(ArrayIncNones, 0)
-    fill_with_none(ptr_dest, m.num_rows_ - last_row);
+        column_data.type().visit_tag([&en, &column_data, &py_dtype, &dest_buffer](auto type_desc_tag) {
+            using TDT = decltype(type_desc_tag);
+            while (auto block = column_data.next<TDT>()) {
+                auto block_pos = 0U;
+                for (auto i = 0U; i < block->row_count(); ++i) {
+                    dest_buffer.cast<PyObject *>(*en) = initialize_array(
+                        py_dtype,
+                        block->shapes() + i,
+                        block->data() + block_pos
+                    );
+                    block_pos += block->shapes()[i];
+                    ++en;
+                }
+            }
+        });
+    } else {
+        column_data.type().visit_tag([&ptr_dest, &column_data, &py_dtype](auto type_desc_tag) {
+            using TDT = decltype(type_desc_tag);
+            while (auto block = column_data.next<TDT>()) {
```

Comment:
Do we have exactly one block per array with >0 dimensional data?

Reply:
No, but the block->row_count() will know how many arrays we have